File: app-android/src/main/java/com/ustadmobile/port/android/view/AddSaleProductToSaleCategoryActivity.java
package com.ustadmobile.port.android.view;

import android.arch.lifecycle.LiveData;
import android.arch.paging.DataSource;
import android.arch.paging.LivePagedListBuilder;
import android.arch.paging.PagedList;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.AddSaleProductToSaleCategoryPresenter;
import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.view.AddSaleProductToSaleCategoryView;
import com.ustadmobile.lib.db.entities.SaleNameWithImage;
import com.ustadmobile.lib.db.entities.SaleProduct;
import com.ustadmobile.port.android.util.UMAndroidUtil;

import java.util.Objects;

import ru.dimorinny.floatingtextbutton.FloatingTextButton;

public class AddSaleProductToSaleCategoryActivity extends UstadBaseActivity
        implements AddSaleProductToSaleCategoryView {

    private Toolbar toolbar;
    private AddSaleProductToSaleCategoryPresenter mPresenter;
    private RecyclerView mRecyclerView;
    private ConstraintLayout addItemCL;
    private TextView addTextView;

    /**
     * This method catches menu buttons/options pressed in the toolbar. Here it is making sure
     * the activity goes back when the back button is pressed.
     *
     * @param item The item selected
     * @return true if accounted for
     */
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        //Setting layout:
        setContentView(R.layout.activity_add_sale_product_to_sale_category);

        //Toolbar:
        toolbar = findViewById(R.id.activity_add_sale_product_to_sale_category_toolbar);
        toolbar.setTitle(getText(R.string.add_item));
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        //RecyclerView
        mRecyclerView = findViewById(
                R.id.activity_add_sale_product_to_sale_category_recyclerview);
        RecyclerView.LayoutManager mRecyclerLayoutManager =
                new LinearLayoutManager(getApplicationContext());
        mRecyclerView.setLayoutManager(mRecyclerLayoutManager);

        addItemCL = findViewById(R.id.activity_add_sale_product_to_sale_category_add_cl);
        addTextView = findViewById(R.id.activity_add_sale_product_to_sale_category_add_text);

        //Call the Presenter
        mPresenter = new AddSaleProductToSaleCategoryPresenter(this,
                UMAndroidUtil.bundleToHashtable(getIntent().getExtras()), this);
        mPresenter.onCreate(UMAndroidUtil.bundleToHashtable(savedInstanceState));


        //Listener
        addItemCL.setOnClickListener(v -> mPresenter.handleAddNewItem());

    }

    /**
     * The DIFF CALLBACK
     */
    public static final DiffUtil.ItemCallback<SaleNameWithImage> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<SaleNameWithImage>() {
                @Override
                public boolean areItemsTheSame(SaleNameWithImage oldItem,
                                               SaleNameWithImage newItem) {
                    return oldItem == newItem;
                }

                @Override
                public boolean areContentsTheSame(SaleNameWithImage oldItem,
                                                  SaleNameWithImage newItem) {
                    return oldItem.equals(newItem);
                }
            };

    @Override
    public void setListProvider(UmProvider<SaleNameWithImage> listProvider) {
        SelectSaleProductToSaleCategoryRecyclerAdapter recyclerAdapter =
                new SelectSaleProductToSaleCategoryRecyclerAdapter(DIFF_CALLBACK, mPresenter, this,
                        false, getApplicationContext());

        // get the provider, set , observe, etc.
        // A warning is expected
        DataSource.Factory<Integer, SaleNameWithImage> factory =
                (DataSource.Factory<Integer, SaleNameWithImage>)
                        listProvider.getProvider();
        LiveData<PagedList<SaleNameWithImage>> data =
                new LivePagedListBuilder<>(factory, 20).build();
        //Observe the data:
        data.observe(this, recyclerAdapter::submitList);

        //set the adapter
        mRecyclerView.setAdapter(recyclerAdapter);
    }

    @Override
    public void setAddtitle(String title) {
        addTextView.setText(title);
    }

    @Override
    public void setToolbarTitle(String title) {
        toolbar.setTitle(title);
    }
}

File: app-android/src/main/kotlin/com/ustadmobile/port/android/view/AddSaleProductToSaleCategoryActivity.kt
package com.ustadmobile.port.android.view

import android.os.Bundle
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.lifecycle.Observer
import androidx.paging.DataSource
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.AddSaleProductToSaleCategoryPresenter
import com.ustadmobile.core.impl.UMAndroidUtil.bundleToMap
import com.ustadmobile.core.view.AddSaleProductToSaleCategoryView
import com.ustadmobile.lib.db.entities.SaleProduct

class AddSaleProductToSaleCategoryActivity
    : UstadBaseActivity(), AddSaleProductToSaleCategoryView {

    private var toolbar: Toolbar? = null
    private var mPresenter: AddSaleProductToSaleCategoryPresenter? = null
    private var mRecyclerView: RecyclerView? = null
    private var addItemCL: ConstraintLayout? = null
    private var addTextView: TextView? = null

    /**
     * This method catches menu buttons/options pressed in the toolbar. Here it is making sure
     * the activity goes back when the back button is pressed.
     *
     * @param item The item selected
     * @return true if accounted for
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        //Setting layout:
        setContentView(R.layout.activity_add_sale_product_to_sale_category)

        //Toolbar:
        toolbar = findViewById(R.id.activity_add_sale_product_to_sale_category_toolbar)
        toolbar!!.title = getText(R.string.add_item)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        //RecyclerView
        mRecyclerView = findViewById(
                R.id.activity_add_sale_product_to_sale_category_recyclerview)
        val mRecyclerLayoutManager = LinearLayoutManager(applicationContext)
        mRecyclerView!!.layoutManager = mRecyclerLayoutManager

        addItemCL = findViewById(R.id.activity_add_sale_product_to_sale_category_add_cl)
        addTextView = findViewById(R.id.activity_add_sale_product_to_sale_category_add_text)

        //Call the Presenter
        mPresenter = AddSaleProductToSaleCategoryPresenter(this,
                bundleToMap(intent.extras), this)
        mPresenter!!.onCreate(bundleToMap(savedInstanceState))


        //Listener
        addItemCL!!.setOnClickListener { v -> mPresenter!!.handleAddNewItem() }

    }

    override fun setListProvider(factory: DataSource.Factory<Int, SaleProduct>) {
        val recyclerAdapter = SelectSaleProductToSaleCategoryRecyclerAdapter(DIFF_CALLBACK, mPresenter!!, this,
                false, applicationContext)

        // get the provider, set , observe, etc.

        val data = LivePagedListBuilder(factory, 20).build()
        //Observe the data:
        data.observe(this,
                Observer<PagedList<SaleProduct>> { recyclerAdapter.submitList(it) })

        //set the adapter
        mRecyclerView!!.adapter = recyclerAdapter
    }

    override fun setAddtitle(title: String) {
        addTextView!!.text = title
    }

    override fun setToolbarTitle(title: String) {
        toolbar!!.title = title
    }

    companion object {

        /**
         * The DIFF CALLBACK
         */
        val DIFF_CALLBACK: DiffUtil.ItemCallback<SaleProduct> = object : DiffUtil.ItemCallback<SaleProduct>() {
            override fun areItemsTheSame(oldItem: SaleProduct,
                                         newItem: SaleProduct): Boolean {
                return oldItem == newItem
            }

            override fun areContentsTheSame(oldItem: SaleProduct,
                                            newItem: SaleProduct): Boolean {
                return oldItem == newItem
            }
        }
    }
}
